packages/backend-storage/src/access_builder.ts

```diff
@@ -0,0 +1,95 @@
+import {
+  AuthRole,
+  ConstructFactory,
+  ConstructFactoryGetInstanceProps,
+  ResourceAccessAcceptor,
+  ResourceAccessAcceptorFactory,
+  ResourceProvider,
+} from '@aws-amplify/plugin-types';
+
+export type StorageAction = 'read' | 'write' | 'delete';
+
+export type StorageAccess = {
+  getResourceAccessAcceptor: (
+    getInstanceProps: ConstructFactoryGetInstanceProps
+  ) => ResourceAccessAcceptor;
+  actions: StorageAction[];
+  resourceSuffix: string;
+};
+
+export type StorageAccessBuilder = {
+  to: (...actions: StorageAction[]) => StorageAccess;
+};
+
+export type EntityAccessBuilder = {
+  authenticated: StorageAccessBuilder;
+  unauthenticated: StorageAccessBuilder;
+  owner: StorageAccessBuilder;
+  resource: (
+    other: ConstructFactory<ResourceProvider & ResourceAccessAcceptorFactory>
+  ) => StorageAccessBuilder;
+};
+
+export const storageAccessBuilder: EntityAccessBuilder = {
```

Comment:
This object can probably be split up a bit for improved readability, but this is the "builder" object that customers would use to define access rules like `allow.owner.to('read', 'write')`.

The output of this pattern is an object with the following properties
1. a callback for resolving the "ResourceAccessAcceptor" for this definition
2. the actions, which more-or-less map directly to IAM actions
3. a suffix (which is * in all cases except for owner rules in which case it's the cognito sub). This is appended to the access path to form the full resource path